File: Schrodinger/schrodinger.cpp
# include <stdio.h>
# include <math.h>
# include <complex>
#include <iostream>

using namespace std;

int main()
{

    //FIJAMOS N, nciclos(max), Lambda y j=0 EL RESTO A UN BUCLE 
    double pi=3.14159265359;
    int N=40;
    int nciclos=N/4;  //hasta N/4
    double l=0.3;   //lambda

    FILE* fichero= fopen("schrodinger_data.dat","a");

    for ( int n=0;n<=nciclos;n++)
    {
        //generamos k0 y s que vayan cambiando para cada n
        double k0=2*pi*n/N;
        double s=1/(4*k0*k0);


        //Bucle que vaya disminuyendo j
        for (int p=0;p==N;p++)
        {
            //generamos Vj
            int j=N-p;
            double Vj;
            if (2*N/5<=j<=3*N/5)
            {
                Vj=l*k0*k0;
            }
            else
            {
                Vj=0;
            }

            //Fi_j,0
            complex<double>c,fijn;
            c=complex<double>(0,k0*j);
            fijn=exp(c)*exp(-8*(4*j-N)*(4*j-N)/(N*N));

        

            if (j==0 and n==0)
            {
                fijn=0;
            }
            if (j==N and n==0)
            {
                fijn=0;
            }

            //cosas de la alfa y la beta
            complex<double> Ajmas,Ajmenos,Aj0,gamma,bjn,a[N],b[N],x;
            Ajmenos=(1,0);
            Ajmas=(1,0);
            Aj0=complex<double> (-2,2/s-Vj);
            
            bjn=4.*fijn/s;

            


            if(j==N-1)
            {
                a[j]=0;
                b[j]=0;
            }

            if(j!=N and j!=N-1)
            {
                gamma=1./(Aj0+Ajmas*a[j+1]);

                a[j]=-Ajmenos*gamma;
                b[j]=gamma*(bjn-Ajmas*bjn);
                
            }

            for (int d=0;d<N;d++)
            {
                x=a[d]*x+b[d];
                fijn=x-fijn;

                //fprintf(fichero,"%f",real(fijn));
                //fprintf(fichero,"\n");
                
            }
            
        }
    
    }


    fclose(fichero);


    return 0;
}

File: Schrodinger/schrodinger.py
import numpy as np
import cmath as cmath
import matplotlib.pyplot as plt
N=1000
nciclos=100
l=0.3 #lambda
n=0
j=0

k0=np.zeros(nciclos)
s=np.zeros(nciclos)

Vj=np.zeros(N,dtype=complex)
fi=np.zeros(N,dtype=complex)

Ajmenos=1
Ajmas=1
Aj0=np.zeros(N,dtype=complex)

a=np.zeros(N,dtype=complex)
a[N-2]=0+0j

gamma=np.zeros(N,dtype=complex)


#generamos k0 y s:
'''
while n < nciclos:
    k0[n]=2*np.pi*(n+1)/N #k0[0] es n=1, k0[1] es n=2...
    s[n]=1/(4*k0[n]**2)
    n=n+1
n=0
'''
k0=2*np.pi*nciclos/N
s=1/(4*k0**2)


#generamos Vj y fi n=0:
while j<N:
    if 2*N/5<=j<=3*N/5:
        Vj[j]=l*k0**2
    else:
        Vj[j]=0

    fi[j]=np.exp(1j*k0*j)*np.exp((-8*(4*j-N)**2)/(N**2))
    j=j+1

fi[0]=0
fi[N-1]=0

norma=np.sum(abs(fi))
#print(norma)
j=0
n=0

#generamos alfa:

for j in range(N):
    Aj0[j]=-2+2j/s-Vj[j]
    


'''
f=N-1
while f>=0:
    gamma[f]=1/(Aj0[f]+a[f])
    a[f-1]=-1*gamma[f]
    a[N-2]=0
    f=f-1 
'''

for i in range(N):
        #alfa
        h=N-i
        if i<N-2:
            a[N-2]=0
            a[0]=0
            gamma[h-1]=1/(Aj0[h-1]+a[h-1])
            a[h-2]=(-1)*gamma[h-1]
            a[N-2]=0
            a[0]=0
                



#calculamos beta, chi, fij+1 
beta=np.zeros(N,dtype=complex)
b=np.zeros(N,dtype=complex)
chi=np.zeros(N,dtype=complex)



n=0

fig,ax=plt.subplots(1,1)

with open("schrodinger_data.dat", "w") as f:

    while n<=N:
        for i in range(N):
            #beta
            h=N-i
            if h<N:
                beta[N-2]=0
                beta[0]=0
                b[h-1]=4j*fi[h-1]/s
                beta[h-2]=gamma[h-1]*(b[h-1]-beta[h-1])
                beta[N-2]=0
                beta[0]=0
        

        for i in range(N):
            #chi
            if i<N-1:
                    #chi[h-1]=a[h-2]*chi[h-2]+beta[h-2]
                chi[0]=0    
                chi[i+1]=a[i]*chi[i]+beta[i]
            chi[N-1]=0

            #fi
                    #    fi[h]=chi[h-1]-fi[h-1]
            fi[i]=chi[i]-fi[i] 
            fi[0]=0
            fi[N-1]=0

            modulo=np.zeros(N,dtype=float)

            fireal=fi.real
            fimag=fi.imag

            modulo=abs(fi)
            norma=np.sqrt(modulo[8])





            f.write(f"{i},{modulo[i]},{Vj[i].real}\n")

        f.write(f"\n")


        linea1=ax.plot(n,norma,'o',mfc='w',mec='k',ms=1)
        
        
    
        
        
        n=n+1

plt.show()

